File: src/main/java/com/github/zametki/model/AbstractId.java
package com.github.zametki.model;

import com.github.mjdbc.type.DbInt;
import org.apache.wicket.util.io.IClusterable;
import org.jetbrains.annotations.NotNull;

/**
 * Base class for all Ids. Serializable DB value.
 */
public class AbstractId implements DbInt, IClusterable, Comparable<AbstractId> {
    public final int intValue;

    public AbstractId(int intValue) {
        this.intValue = intValue;
    }

    @Override
    public int getDbValue() {
        return intValue;
    }


    public int hashCode() {
        return intValue;
    }

    @Override
    public final boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        AbstractId that = (AbstractId) o;
        return intValue == that.intValue;
    }

    public String toString() {
        return getClass().getSimpleName() + "[" + intValue + "]";
    }

    public boolean isValid() {
        return intValue > 0;
    }
    public boolean isInvalid() {
        return !isValid();
    }

    public boolean isRoot() {
        return !isValid();
    }
    @Override
    public int compareTo(@NotNull AbstractId o) {
        return Integer.compare(intValue, o.intValue);
    }
}

File: src/main/java/com/github/zametki/model/AbstractId.kt
package com.github.zametki.model

import com.github.mjdbc.type.DbInt
import org.apache.wicket.util.io.IClusterable

/**
 * Base class for all Ids. Serializable DB value.
 */
open class AbstractId(val intValue: Int) : DbInt, IClusterable, Comparable<AbstractId> {

    val isValid: Boolean get() = intValue > 0
    val isInvalid: Boolean get() = !isValid
    val isRoot: Boolean get() = !isValid

    override fun getDbValue() = intValue

    override fun hashCode() = intValue

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as AbstractId?
        return intValue == that!!.intValue
    }

    override fun toString() = javaClass.simpleName + "[" + intValue + "]"

    override fun compareTo(other: AbstractId) = Integer.compare(intValue, other.intValue)
}
